File: presentation/src/main/java/com/balv/imdb/ui/home/listview/HomePagingAdapter.java
package com.balv.imdb.ui.home.listview;

import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.paging.PagingDataAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.balv.imdb.R;
import com.balv.imdb.domain.models.Movie;
import com.balv.imdb.ui.home.listener.IHomeItemClickListener;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.bitmap.CenterCrop;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;
import com.bumptech.glide.request.RequestOptions;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import dagger.hilt.android.scopes.FragmentScoped;

@FragmentScoped
public class HomePagingAdapter extends PagingDataAdapter<Movie, HomePagingAdapter.ItemViewHolder> {
    private static final String TAG = "HomeAdapter";
    private List<Movie> mMovieList = new ArrayList<>();
    @Inject
    public IHomeItemClickListener mClickListener;

    @Inject
    public HomePagingAdapter(){
        super(DIFF_CALLBACK);
    }

    private static final DiffUtil.ItemCallback<Movie> DIFF_CALLBACK = new DiffUtil.ItemCallback<Movie>() {
        @Override
        public boolean areItemsTheSame(@NonNull Movie oldItem, @NonNull Movie newItem) {
            return TextUtils.equals(oldItem.getId(), newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull Movie oldItem, @NonNull Movie newItem) {
            return TextUtils.equals(oldItem.getImdbRated(), newItem.getImdbRated());
        }
    };


    public void setDataList(List<Movie> movieList) {
        mMovieList = movieList;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public HomePagingAdapter.ItemViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.home_item, parent, false);
        return new ItemViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(HomePagingAdapter.ItemViewHolder holder, int position) {
        holder.bind(getItem(position));
    }


    class ItemViewHolder extends RecyclerView.ViewHolder {
        View view;
        TextView title;
        TextView rating;
        TextView date;
        ImageView thumb;

        public ItemViewHolder(View itemView) {
            super(itemView);
            view = itemView;
            title = view.findViewById(R.id.movie_title);
            rating = view.findViewById(R.id.imdb_rating);
            date = view.findViewById(R.id.release_date);
            thumb = view.findViewById(R.id.thumb);
        }

        public void bind(Movie data){
            if (data != null) {
                view.setOnClickListener(view -> mClickListener.onItemClick(data.getId()));
                title.setText(data.getTitle());
                rating.setText(data.getImdbRated());
                date.setText(data.getReleased());
                RequestOptions requestOptions = new RequestOptions();
                requestOptions = requestOptions.transforms(new CenterCrop(), new RoundedCorners(16));
                Glide
                        .with(view.getContext())
                        .load(data.getPoster())
                        .centerCrop()
                        .placeholder(R.drawable.mv_place_holder)
                        .apply(requestOptions)
                        .into(thumb);
            }
        }
    }
}

File: app/src/main/java/com/balv/imdb/ui/home/listview/HomePagingAdapter.kt
package com.balv.imdb.ui.home.listview

import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.balv.imdb.R
import com.balv.imdb.domain.models.Movie
import com.balv.imdb.ui.home.listener.IHomeItemClickListener
import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.bitmap.CenterCrop
import com.bumptech.glide.load.resource.bitmap.RoundedCorners
import com.bumptech.glide.request.RequestOptions
import dagger.hilt.android.scopes.FragmentScoped
import javax.inject.Inject

@FragmentScoped
class HomePagingAdapter @Inject constructor() :
    PagingDataAdapter<Movie, HomePagingAdapter.ItemViewHolder>(
        DIFF_CALLBACK
    ) {
    private var mMovieList: List<Movie> = ArrayList()

    @JvmField
    @Inject
    var mClickListener: IHomeItemClickListener? = null

    fun setDataList(movieList: List<Movie>) {
        mMovieList = movieList
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
        val itemView =
            LayoutInflater.from(parent.context).inflate(R.layout.home_item, parent, false)
        return ItemViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
        holder.bind(getItem(position))
    }


    inner class ItemViewHolder(var view: View) : RecyclerView.ViewHolder(view) {
        var title: TextView = view.findViewById(R.id.movie_title)
        var rating: TextView = view.findViewById(R.id.imdb_rating)
        var date: TextView = view.findViewById(R.id.release_date)
        var thumb: ImageView =
            view.findViewById(R.id.thumb)

        fun bind(data: Movie?) {
            if (data != null) {
                view.setOnClickListener { view: View? ->
                    mClickListener!!.onItemClick(
                        data.id
                    )
                }
                title.text = data.title
                rating.text = data.voteAverage.toString()
                date.text = data.releaseDate
                var requestOptions = RequestOptions()
                requestOptions = requestOptions.transforms(CenterCrop(), RoundedCorners(16))
                Glide
                    .with(view.context)
                    .load(data.posterPath)
                    .centerCrop()
                    .placeholder(R.drawable.mv_place_holder)
                    .apply(requestOptions)
                    .into(thumb)
            }
        }
    }

    companion object {
        private const val TAG = "HomeAdapter"
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<Movie> =
            object : DiffUtil.ItemCallback<Movie>() {
                override fun areItemsTheSame(oldItem: Movie, newItem: Movie): Boolean {
                    return oldItem.id == newItem.id
                }

                override fun areContentsTheSame(oldItem: Movie, newItem: Movie): Boolean {
                    return oldItem.voteAverage == newItem.voteAverage
                }
            }
    }
}
